### PR TITLE
Implements our StyleGuide

### PR DESCRIPTION
**Pull request by jibees**
#### What? Why?
This PR is here to create a StyleGuide inside OFN using `view_component` and `storybook`

##### How it works?
 - Components should now be created with https://viewcomponent.org with helper method: `rails generate component Example title --sidecar` (see document for more details)
 - Stories should be put into the `spec/components/stories` folder and then transformed to be ready for Storybook with the command: `rake view_component_storybook:write_stories_json`. This creates a `example_component.stories.json`, which is gitignored, but readable by StoryBook.
 - Once this is done, we have to start the rails server, and then StoryBook itself with the following command: `yarn storybook`
 - Enjoy!

Source:
 - https://github.com/github/view_component
 - https://github.com/jonspalmer/view_component_storybook
 - https://github.com/storybookjs/storybook

#### Release notes
Create our styleguide

Changelog Category: Technical changes

#### What should we test?

- Test the _Print to PDF_ functionality everywhere in the app.
- Test CSS and fonts look okay
- Check the enterprise title renders okay in shopfronts

##### MonkeyPatch
For now, this is not functional. Indeed, `view_component` is using a monkeypatch to override `#render` method since we're using Rails5.0 (seems to be ok with Rails5.2, but not sure): https://github.com/github/view_component/blob/main/lib/view_component/render_monkey_patch.rb
As `WickedPdf` used the same monkeypatch, seems to be a conflict, as noticed here: https://github.com/github/view_component/issues/288
Not sure to totally understand, still need some investigation.
For the record, the error in the rails console: 
```
....
wicked_pdf (2.1.0) lib/wicked_pdf/pdf_helper.rb:46:in `render_with_wicked_pdf'
wicked_pdf (2.1.0) lib/wicked_pdf/pdf_helper.rb:30:in `render'
wicked_pdf (2.1.0) lib/wicked_pdf/pdf_helper.rb:46:in `call'
wicked_pdf (2.1.0) lib/wicked_pdf/pdf_helper.rb:46:in `render_with_wicked_pdf'
wicked_pdf (2.1.0) lib/wicked_pdf/pdf_helper.rb:30:in `render'
wicked_pdf (2.1.0) lib/wicked_pdf/pdf_helper.rb:46:in `call'
....
```
Resolved by cherrypicked: 2b19ea9a5bab0ad4847202503050f7cb02afabd3 and 3893c0778534a55848f3c9e7d738ab841a82a41e

### CODE REVIEW

**Comment by luisramos0** [comment]
This is nice :clap:
I wonder why so many changes in yarn.lock.

Could you share a screenshot of storybook running?

**Comment by Matt-Yorkley** [comment]
Okay, option 1 is to use the ViewComponents config option `render_monkey_patch_enabled = false`, and then render components in the app using the `#render_component` method. This should work fine, but I think it won't work with StoryBook for now.

Option 2 might get a bit messy, we'll have to add an additional monkey-patch to wicked_pdf ourselves. We might already have a problem here, in that the current version already uses `alias_method_chain`, which is removed in Rails 5.2... there's definitely some complexity there.

**Comment by jibees** [comment]
Thanks a lot @Matt-Yorkley .
I've cherrypicked your two commits, and now the error disappeared. 

Now I'm facing over CORS errors: 
```
iframe.html?id=example-component--with-short-text&viewMode=story:1 Access to fetch at 'http://localhost:3000/rails/stories/example_component/with_short_text?title=OK' from origin 'http://localhost:61898' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```
Capture d’écran 2021-03-31 à 21 07 50

**Comment by jibees** [comment]
Here you go!
Capture d’écran 2021-03-31 à 21 25 45

**Comment by Matt-Yorkley** [comment]
> I've cherrypicked your two commits, and now the error disappeared

It's possible those two commit might break PDFs elsewhere in the app, it was a quick fix, but it would need careful testing...

**Comment by jibees** [comment]
> > I've cherrypicked your two commits, and now the error disappeared
> 
> It's possible those two commit might break PDFs elsewhere in the app, it was a quick fix, but it would need careful testing...

I already tried to make print to pdf functionality to work on my machine, but i always failed... ;(

**Comment by Matt-Yorkley** [comment]
Hot tip: `spec/services/invoice_renderer_spec.rb` needs a little tweak on **line 38**, like this:
```diff
--  .to receive(:render_to_string)
++  .to receive(:render_to_string_with_wicked_pdf)
```

**Comment by jibees** [comment]
> Hot tip: `spec/services/invoice_renderer_spec.rb` needs a little tweak on **line 38**, like this:
> 
> ```diff
> --  .to receive(:render_to_string)
> ++  .to receive(:render_to_string_with_wicked_pdf)
> ```

Thanks. I squashed your commit. 🙏

File: package.json
Comment (Matt-Yorkley) [inline package.json:22]: Shouldn't this be under `devDependencies`? :thinking:

File: Gemfile
Comment (Matt-Yorkley) [inline Gemfile:116]: I'd prefer a different solution here, if there is one.

The problem was that loading StoryBook in dev was throwing CORS errors, right? I think we can probably solve that without adding a new gem.

File: package.json
Comment (jibees) [reply package.json:22]: Actually, it's not that easy, because: 
 - `dependencies`: Packages required by your application in production.
 - `devDependencies`: Packages that are only needed for local development and testing.

StoryBook can be run in production, to show the components used, but is clearly used for development.
I don't have a strong opinion on this. I will do what you think is more logical with the OFN standard.

File: Gemfile
Comment (jibees) [reply Gemfile:116]: > The problem was that loading StoryBook in dev was throwing CORS errors, right?

Exactly. You can see the error in this comment: https://github.com/openfoodfoundation/openfoodnetwork/pull/7284#issuecomment-811356981

Just to understand: what is the main problem by adding some gems? Are you concerned about the systematic increase in the number of dependencies, which can be annoying at some times?

File: package.json
Comment (Matt-Yorkley) [reply package.json:22]: I think we don't want to load all this extra code in production, if possible :+1:

File: Gemfile
Comment (Matt-Yorkley) [reply Gemfile:116]: > Are you concerned about the systematic increase in the number of dependencies, which can be annoying at some times?

Yeah, we should try to keep dependencies to a minimum if possible :+1:

Comment (jibees) [reply Gemfile:116]: I think I have reached the limits of my skills with rails: I can use a `gem`, but replace it with our code seems to be (at the moment) too difficult for me. 
I'd really appreciate if you could help me on this one. 🙏

Comment (jibees) [reply Gemfile:116]: (BTW, I can understand the concern about having an eye on the number of dependencies. On the other hand, I also concerned about the NIH syndrome, and I think each dependencies should be looked by itself and without being too categorical.)

Comment (mkllnk) [reply Gemfile:116]: Matt is not suggesting to add a complex solution to our codebase that a gem could do better. He suggests that we try to understand why our codebase has an issue with iframes and then the solution may be very simple. A bit of OFN history knowledge helps here.

We started restricting iframes when we offered embedded shopfronts. They never worked properly and the feature is deprecated now. But the restriction is still there:
https://github.com/openfoodfoundation/openfoodnetwork/blob/2128675e955cc1f1cbae875bbac47dac0e8aafff/app/controllers/application_controller.rb#L93-L102

Maybe we can add a rule there to allow requests for Storybook.

Comment (jibees) [reply Gemfile:116]: It's not exactly related to iframe, but more about CORS policy.
But anyway, will look to `openfoodnetwork/app/controllers/application_controller.rb` thanks for pointing me that out!

File: app/controllers/stories_controller_decorator.rb
Comment (jibees) [inline app/controllers/stories_controller_decorator.rb:1]: @mkllnk @Matt-Yorkley 
I hope this solution sounds good to you. 👍

File: Gemfile
Comment (jibees) [reply Gemfile]: See commit 064bc7025aab4445d78e4e6be821519a41e1ec42 : I've wrote a decorator for the Controller that serves the resources and sets the right CORS policy. Hope it's good now!

File: app/controllers/stories_controller_decorator.rb
Comment (mkllnk) [inline app/controllers/stories_controller_decorator.rb:8]: Great approach!

I'm just wondering if we should limit this to the development environment.

The extension can also be done in a cleaner way. You can define a new module with a method: `PermissiveCORSHeaders.before`. Then you can add this line in an initializer:
```rb
ViewComponent::Storybook::StoriesController.before_filter(PermissiveCORSHeaders) if Rails.env.development?
```

Examples:

- Look at `config/initializers/paypal.rb`.
- Search our code for `CurrentOrderLocker`.

**Review by mkllnk — COMMENTED** [review]
Great work! I really appreciate how you take every challenge we through at you and learn along the way. You are making excellent progress.

I added some questions for my understanding. And do you think it would be possible to create a real example that we can use? Even if it's really small, like a text box somewhere.

File: Gemfile
Comment (jibees) [reply Gemfile:116]: Since ViewComponent will be used to render the component in both the development and production environment, I guess not. But maybe I'm wrong about this group concept. 
See exemple from Github PrimerViewComponents: https://github.com/primer/view_components/blob/740cd4821e544bbcdca5f1d65f07468cde8457a1/Gemfile#L36

File: app/components/example_component.rb
Comment (jibees) [reply app/components/example_component.rb:3]: Actually, command `rails generate component Example title --sidecar` creates the spec for the component (see https://viewcomponent.org/guide/#quick-start) but I deleted it since I had no idea which kind of content I should put into (it's empty by default). 
Will recreate it, and try to make it useful.

Comment (jibees) [reply app/components/example_component.rb:3]: Some example done by commit: 00b92a51e
Very simple at the moment, but it's a start, and only an example.

File: config/application.rb
Comment (jibees) [reply config/application.rb:9]: it's almost the same question that we had already: do we need Storybook in production? I guess somedays, yes (at least why not in certain specific cases), but at this time, I think it's more a developper tools so I will make it only for development, and re-create a PR if we want to push StoryBook in production.

Comment (jibees) [reply config/application.rb]: It's now in the `config/environments/development.rb` file: 

https://github.com/openfoodfoundation/openfoodnetwork/pull/7284/files#diff-d3c4b3f41072daa416f1920511e9b2e26caea8c5cec0a14cb9508589a4dafa47R1-R2

File: config/initializers/wicked_pdf.rb
Comment (jibees) [reply config/initializers/wicked_pdf.rb:13]: I let @Matt-Yorkley answer on this point ; he helps me a lot on this particular point... 🙏

File: app/controllers/stories_controller_decorator.rb
Comment (jibees) [reply app/controllers/stories_controller_decorator.rb:8]: > I'm just wondering if we should limit this to the development environment.

Yes, I think as we know assume that StoryBook will only run for the development env. it's a great idea.

File: config/application.rb
Comment (jibees) [reply config/application.rb]: I guess I didn't understand your point, because if I put this one under `config/environments/development.rb` got the error:

```
uninitialized constant ViewComponent::Storybook::StoriesController (NameError)
```

Reverted, and put into `config/application.rb`

File: app/controllers/stories_controller_decorator.rb
Comment (jibees) [reply app/controllers/stories_controller_decorator.rb]: There you go: b7399f43fe153efb71e36b38133691297d57ad0b !

**Comment by jibees** [comment]
> Great work! I really appreciate how you take every challenge we through at you and learn along the way. You are making excellent progress.

Thanks a lot for your Incentives, really appreciate. I must confess it's sometimes a little bit disappointed these _incessant back and forth_ and I'm sometimes frustrating about how many times it takes me to develop new feature in this stack. But anyway, yeah, I learn a lot, and maintaining code quality at this level requires this drawbacks. 

> And do you think it would be possible to create a real example that we can use? Even if it's really small, like a text box somewhere.

Will add a real & simple example inside our codebase yes!

File: app/controllers/stories_controller_decorator.rb
Comment (jibees) [reply app/controllers/stories_controller_decorator.rb]: It's very strange, because I _sometimes_ had CORS error, and I need to restart the sever. Don't know why.

```
Access to fetch at 'http://localhost:3000/rails/stories/distributor_title_component/default?name=Freddy+s+Farm+Shop' from origin 'http://localhost:64504' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

File: Gemfile
Comment (Matt-Yorkley) [reply Gemfile:116]: Yep, it should stay there :+1: It'll be used in all environments.

File: app/components/example_component.rb
Comment (Matt-Yorkley) [reply app/components/example_component.rb:3]: We'll be able to start adding really great unit tests for all UI elements as we move to rendering with `ViewComponent`.

Yay :tada::partying_face:

**Comment by jibees** [comment]
same_picture

File: config/initializers/wicked_pdf.rb
Comment (Matt-Yorkley) [reply config/initializers/wicked_pdf.rb:13]: > It would be nice to mention where you found this

I wrote it. Basically _both_ `wicked_pdf` and `view_component` apply a slightly different monkey-patch on the `ActionController#render` and `ActionController#render_to_string` methods, but they're not compatible. I wrote an initial comment here: https://github.com/openfoodfoundation/openfoodnetwork/pull/7284#issuecomment-811211809

So this basically removes the `wicked_pdf` monkey-patch for those methods, which means when we're rendering a PDF we now need to explicitly use `#render_to_pdf_with_wicked_pdf` (for example).

Yeah it's a bit of a mess... :see_no_evil: 

Maybe we can add some more detailed comments to make the context a bit clearer.

File: config/application.rb
Comment (Matt-Yorkley) [reply config/application.rb:9]: I agree with @mkllnk, it feels like we probably shouldn't be loading this engine in `application.rb`.

I'll take a look at it now, and see what we can do :ok_hand:

Comment (Matt-Yorkley) [reply config/application.rb:9]: Looks like it's time to `bundle install && yarn install` and RTFM :sweat_smile:

File: app/components/example_component.rb
Comment (Matt-Yorkley) [reply app/components/example_component.rb:3]: And that means we can start kicking a lot of stuff further down the testing pyramid :muscle:

**Comment by Matt-Yorkley** [comment]
@jibees I submitted a PR to your branch here: https://github.com/jibees/openfoodnetwork/pull/1

After playing with Storybook, it looks like there are a few minor issues:

- If I add that missing apostrophe in `Freddy's Farm Shop` for the example component and re-run the rake task and reload everything, it actually seems to break the Story..? :confused: 
- There's a message in the UI saying `No code available` when viewing a component under the `Docs` tab. Is it supposed to be showing the source code there?
- I get some console errors related to the CSS when running Storybook. Maybe it's not a big deal, it seems to work.
- When I boot the _main app_ in dev, I now get some issues where the icon font is missing, with related CORS errors. I think we need to look at the CORS thing again and tweak it a bit, but I haven't looked at it yet.

**Comment by jibees** [comment]
> @jibees I submitted a PR to your branch here: [jibees#1](https://github.com/jibees/openfoodnetwork/pull/1) to address some of those engine loading / initializer comments 👍

Thanks a lot for that. 🙏 
Should I squash your commits on my branch? To have a (maybe) cleaner final PR?


> 
> After playing with Storybook, it looks like there are a few minor issues:
> 
> * If I add that missing apostrophe in `Freddy's Farm Shop` for the example component and re-run the rake task and reload everything, it actually seems to break the Story..? 😕

Yes, at some point (I didn't investigate too much) there is a bad single quote escape (🙈 ...)


> * There's a message in the UI saying `No code available` when viewing a component under the `Docs` tab. Is it supposed to be showing the source code there?

Yes. It's done with `@storybook/addon-docs`. No idea if it's compatible with view_components...

> * I get some console errors related to the CSS when running Storybook. Maybe it's not a big deal, it seems to work.

Yes I saw this. It's because the `iframe` first looks for (in `.storybook/preview-head.html`): 
```html
<link rel="stylesheet" media="screen" href="${SERVER_URL}/assets/darkswarm/all.css" />
```
but then, `${SERVER_URL}` is replaced (by `http://localhost:3000`) thanks to `.storybook/preview.js`

I had, at this time, no better idea.

> * When I boot the _main app_ in dev (not Storybook), I now get some display issues where the icon font is missing, with related CORS errors. I think we need to look at the CORS thing again and tweak it a bit, but I haven't looked at it yet.

Strange... Actually, as Strorybook needs to access to fonts and CSS, it also needs to configure the CORS policies for `/assets/` route.

**Comment by Matt-Yorkley** [comment]
Alright, I debugged the broken font issue a bit. It looks like this line changed the default behavior:
```ruby
config.action_controller.asset_host =  'http://localhost:3000'
```

I usually access my dev app by putting `0.0.0.0:3000` in the browser, but now that we're explicitly setting the asset host to `localhost:3000`, it's broken. Putting `localhost:3000` in the browser seems to work fine (the fonts are fixed), so I'm happy to go with that unless anyone else objects :ok_hand:

On the rendering single quotes issue: it seems to be a known bug in the `@storybook/addon-controls` package. Possibly applies to other HTML entities as well... :man_shrugging:

File: Gemfile
Comment (mkllnk) [reply Gemfile:116]: Ah, yes, that sounds right! Storybook should go into development.

File: config/initializers/stories.rb
Comment (mkllnk) [inline config/initializers/stories.rb:2]: This empty line is superfluous.

File: app/controllers/stories_controller_decorator.rb
Comment (mkllnk) [reply app/controllers/stories_controller_decorator.rb]: Okay, let's monitor that and see if it's a problem. Maybe there was just a caching problem on your machine.

File: app/components/distributor_title_component/distributor_title_component.html.haml
Comment (mkllnk) [inline app/components/distributor_title_component/distributor_title_component.html.haml:1]: I've never seen this syntax. Thanks for pointing me to it. This example is slightly simpler like this:
```suggestion
%h3= @name
```
The `=` syntax is also a bit more robust for Ruby code.

> Haml uses an overly simplistic regular expression to identify string interpolation rather than a full-blown Ruby parser. This is fast and works for most code but you may have errors...
> https://haml.info/docs/yardoc/file.REFERENCE.html#gotchas

File: spec/components/stories/distributor_title_component_stories.rb
Comment (mkllnk) [inline spec/components/stories/distributor_title_component_stories.rb:7]: Superfluous empty line.

File: .storybook/preview-head.html
Comment (mkllnk) [inline .storybook/preview-head.html:2]: This doesn't work?
```suggestion
<link rel="stylesheet" media="screen" href="/assets/darkswarm/all.css" />
```

Comment (mkllnk) [reply .storybook/preview-head.html:2]: When we re-design the backoffice, it would be good to make it compatible with the frontend styles, maybe even build on top of them. Then we can include them both here but normal shoppers will need only the frontend styles.

File: app/views/shopping_shared/_header.html.haml
Comment (mkllnk) [inline app/views/shopping_shared/_header.html.haml:9]: This is cool! :sunglasses:

**Review by mkllnk — COMMENTED** [review]
This is getting better and better. :smile: 

Introducing new technology can be difficult. This one doesn't seem to be completely straight forward but we are getting there. Good team work!

> Should I squash your commits on my branch? To have a (maybe) cleaner final PR?

When a PR has been reviewed already, as Matt and I did with yours here, it's much easier for us if you don't squash the commits. We can then see what you actually changed compared to your previous version and if that addressed our concerns. Once we approved and you want to put the PR further forward, it's good to squash. The cleaner PR will be easier to review for Pau, for example, who hasn't seen the work in progress yet.

File: .storybook/preview-head.html
Comment (jibees) [reply .storybook/preview-head.html:2]: > This doesn't work?

Actually not, because it's then served by the Storybook server (ie. `http://localhost:52533/assets/darkswarm/all.css`). We need to specify the Rails server to get the right assets. 

> When we re-design the backoffice, it would be good to make it compatible with the frontend styles, maybe even build on top of them. Then we can include them both here but normal shoppers will need only the frontend styles.

That would be a good idea. I've tested (a little bit) to share CSS between back and front office (for the unit price tooltip) and it was a bit painful. But maybe this is not awful as I saw. Worth a try!

**Comment by jibees** [comment]
> I think we can delete that line and everything will work fine 👍

I deleted this useless line.

**Comment by jibees** [comment]
@Matt-Yorkley I merged your branch onto mine (with the github interface), but I don't know if I shouldn't have cherrypicked your commits instead. What do you think about it?

**Comment by jibees** [comment]
@Matt-Yorkley 
> On the rendering single quotes issue: it seems to be a known bug in the `@storybook/addon-controls` package. Possibly applies to other HTML entities as well... 🤷‍♂️

As far as I see, it's a bug relative to the webpack loader used to load to stories in `json` format. 
Error: 

Capture d’écran 2021-04-12 à 10 29 02

The loader used to load stories in `json` from the webpack configuration:
```js
      {
        type: 'javascript/auto',
        test: /\.stories\.json$/,
        use: '/Users/jibees/dev/openfoodnetwork/node_modules/@storybook/server/dist/cjs/server/loader.js'
      },
```

**Comment by jibees** [comment]
> @Matt-Yorkley I merged your branch onto mine (with the github interface), but I don't know if I shouldn't have cherrypicked your commits instead (since I cannot perform an interactive rebase now). What do you think about it?

Forget my comment, I now have one branch, with your commits.

File: .storybook/preview-head.html
Comment (mkllnk) [reply .storybook/preview-head.html:2]: Interesting. I expected the path to be relative to the host from which it's loaded but Storybook works differently here. I'm not sure if it's the fact that the element is injected via JS or maybe it's Storybook injecting the code into the page instead of just using an iframe. :shrug: 

Anyway. The current solution doesn't seem to work either. The variable doesn't get replaced in time and my browser tries to load the raw URL. I'll add a commit with another solution.

File: .storybook/preview-head.html
Comment (jibees) [inline .storybook/preview-head.html:2]: 👌

File: .storybook/preview-head.html
Comment (jibees) [reply .storybook/preview-head.html]: Thanks.

**Comment by jibees** [comment]
Dear @openfoodfoundation/core-devs 
I think we're almost here, don't you?
There's still an issue around `@storybook/server` and single quote escaping thought the stories in json format, that might be investigate (but this PR can be merged without)
It also needs a bit of clean history, to have a proper git history: once it's done I think it's worth having this PR in `Code Review`.
What do you think?

**Comment by jibees** [comment]
FYI, I've also open an issue into the `view_component_storybook` project, on how to handle single quote escaping issue : 
https://github.com/jonspalmer/view_component_storybook/issues/31

I'm facing the same problem with I use their example: https://github.com/jonspalmer/view_component_storybook_example

File: config/initializers/storybook.rb
Comment (Matt-Yorkley) [inline config/initializers/storybook.rb:8]: I think after the 5.2 upgrade is merged we could even replace this with something slightly nicer, as 5.2 introduces a new built-in DSL for modifying response headers :tada:

**Review by Matt-Yorkley — APPROVED** [review]

**Review by mkllnk — APPROVED** [review]

**Review by sauloperez — APPROVED** [review]
Awesome @jibees this is really good stuff! :clap: I :heart: the tech choice too.

I have a question though. Is this the granularity level of components that we aim to build with `ViewComponent`? I'm asking because obviously tons of more Ruby objects add up when rendering views, which tends to be particularly slow already.

**Comment by Matt-Yorkley** [comment]
> Is this the granularity level of components that we aim to build with ViewComponent?

Probably not as granular as the example...

> Ruby objects will add up when rendering views, which tends to be particularly slow already

The good news there is that rendering via `ViewComponent` is 10 times faster than regular Rails partial rendering.

**Comment by sauloperez** [comment]
> The good news there is that rendering via ViewComponent is 10 times faster than regular Rails partial rendering.

Hey hey hey! I really like that!

I played with another framework like this one on another project and complex card components where a bit slow to render, that's why I was a bit concerned.

**Comment by jibees** [comment]
> Probably not as granular as the example...

That's my tropism as a React developper : everything is a component 🙈  But well, for an `h2`, it seems to be totally useless ;)

> The good news there is that rendering via `ViewComponent` is 10 times faster than regular Rails partial rendering.

<3

> We'll need to document a bit things like how to run it, how to approach stories, how to run it with docker (we might need a storybook container in our docker-compose.yml) etc. I trust we'll do that once we start using it.

I can start a wiki page!

**Comment by sauloperez** [comment]
> I can start a wiki page!

that would be ace! Just having the page there with a paragraph will be enough to iterate on it as we move along :ok_hand:

**Comment by jibees** [comment]
@Matt-Yorkley 
> On the rendering single quotes issue: it seems to be a known bug in the `@storybook/addon-controls` package. Possibly applies to other HTML entities as well... 🤷‍♂️

Fixed in `storybook` 6.2.9: https://github.com/storybookjs/storybook/pull/14615#issuecomment-821029341 !

**Comment by filipefurtad0** [comment]
Hey @jibees
Congratulations on this work! :clap:  :clap:  :clap: 

Performed these tests in my system (Ubuntu/Firefox) and briefly checked mobile (iOS) as well.

**Tested both invoice types**, as set in `/admin/invoice_settings/edit` (superadmin):
- regular invoice model
- alternative invoice model (includes enterprise logo)

and created pdf files under (results reflect tests on both pdf types):
i) `/orders`: ticking one order and clicking "Print Invoices" -> print 1 pdf - OK
ii) `/orders`: ticking two or more orders and clicking "Print Invoices" -> printed a multi-page pdf - OK
iii) `/admin/orders/<order_number>/edit` - under the Actions menu, select "Print Invoice" -> `link admin/orders/<order_number>/print` **not rendering pdf correctly**:

Opens a new tab but displays:
image

- `/admin/orders/<order_number>/edit` - under the Actions menu, select "Send Invoice" -> sends a pdf as en Email attachment - OK

iv) T&Cs pdf upload: as an admin, under `admin/enterprises/<enterprise_name>/edit#/business_details` - OK
v)  T&Cs pdf view: as a customer, on checkout - OK


**Test CSS and fonts** look okay

Looked into:
- superadmin pages
- admin pages
- customer pages
- pages which require no log-in (homepage, map, producers, etc)

Found no issues. (The exception is the `/print` page on the test case v)  above, in which formatting breaks). 

**Checked the enterprise title** renders okay in shopfronts - the short video below switches tabs between before and after deploy:
Peek 2021-04-22 11-12

I can confirm "they're the same picture" :-D 
The enterprise title renders as before in shopfronts, we're good here.

**Summary**
So, we're really almost done here - just point v) needs to be addressed. I'm not sure if this means much more code; I'm moving back to in dev but please feel free to move it back to test ready @jibees if you feel is does not require additional code review.

PS: it seems `view_component` this seems to introduce some very interesting features in terms of testing, both as unit or feature tests:
https://viewcomponent.org/guide/testing.html 
Something to have fun with in the future :star_struck:

**Comment by Matt-Yorkley** [comment]
Okay, `app/controllers/spree/admin/orders_controller.rb` line 92 needs this change:
```diff
-- render InvoiceRenderer.new.args(@order)
++ render_with_wicked_pdf InvoiceRenderer.new.args(@order)
```
and this also indicates we have no tests on that PDF print button? We can make a separate issue to cover it :+1:

Are there any cases where we send PDFs attached to emails?

**Comment by filipefurtad0** [comment]
Thanks for feedback @Matt-Yorkley ! :+1: 
Created two tech-debt issues to cover:

- the missing test case on pdf/invoice creation - #7455
- email with attached pdf/invoice - #7456

**Comment by jibees** [comment]
@sauloperez @Matt-Yorkley 
Draft: https://github.com/openfoodfoundation/openfoodnetwork/wiki/Creating-new-components Feel free to comment/edit/whatever.
Thanks a lot guys!

**Comment by jibees** [comment]
@filipefurtad0 @Matt-Yorkley 
Thanks a lot for your testing, and for issues open.
I can do a test, but unfortunately, I have never been able to print to PDF on my machine. I still have a problem around `wicked_pdf` ... @Matt-Yorkley can you please take a look at these two problems?

**Comment by Matt-Yorkley** [comment]
> I have never been able to print to PDF on my machine.

Me neither, but I think we recently merged a change to master that means we can now run those PDF tests locally. I haven't re-checked it yet though.

As long as it passes manual testing here, I think we can cover those test coverage issues separately, without blocking this PR :ok_hand:

**Comment by jibees** [comment]
> As long as it passes manual testing here, I think we can cover those test coverage issues separately, without blocking this PR 👌

Yes, you're right, I agree. I remove dependencies.

**Comment by jibees** [comment]
@Matt-Yorkley 
Modification done: 9a670c782106339d1ef1be96e8cbb17828cfca02 Thanks a lot for this one! 🙏 
Moving back to `Code Review`

**Review by Matt-Yorkley — APPROVED** [review]

**Comment by filipefurtad0** [comment]
Alright!

After your changes, clicking "Print Invoice" on the Actions menu or visiting `/admin/orders/<order_nr>/print` now prints the respective pdf invoice, for **both invoice** types.

Can't think of anywhere else where pdfs are created, I think we're good to go now.

Thanks for the quick fix :rocket:

**Comment by Matt-Yorkley** [comment]
Alright, lets slap the button.

Thanks @jibees, sorry it was such a convoluted journey! :sweat_smile:

**Pull request merged**
